File: server/src/weather-spatial-index.js
import fs from 'node:fs'
import path from 'node:path'

export class WeatherSpatialError extends Error {
  constructor(kind, message) { super(message); this.name = 'WeatherSpatialError'; this.kind = kind }
}

const LEVELS = ['province', 'city', 'county', 'township', 'village']
function fail(message = '天气空间数据不可用') { throw new WeatherSpatialError('unconfigured', message) }
function readJson(file) { try { return JSON.parse(fs.readFileSync(file, 'utf8')) } catch { fail() } }
function safePath(root, relative) {
  if (typeof relative !== 'string' || !relative || path.isAbsolute(relative) || relative.split(/[\\/]/).includes('..')) fail('天气边界引用无效')
  const target = path.resolve(root, relative), base = path.resolve(root)
  if (target !== base && !target.startsWith(`${base}${path.sep}`)) fail('天气边界引用越界')
  return target
}
function finitePoint(point) { return Array.isArray(point) && point.length === 2 && point.every(Number.isFinite) }
function validRing(ring) { return Array.isArray(ring) && ring.length >= 4 && ring.every(finitePoint) && ring[0][0] === ring.at(-1)[0] && ring[0][1] === ring.at(-1)[1] }
function validGeometry(geometry) {
  if (geometry?.type === 'Polygon') return Array.isArray(geometry.coordinates) && geometry.coordinates.length > 0 && geometry.coordinates.every(validRing)
  if (geometry?.type === 'MultiPolygon') return Array.isArray(geometry.coordinates) && geometry.coordinates.length > 0 && geometry.coordinates.every((polygon) => Array.isArray(polygon) && polygon.length > 0 && polygon.every(validRing))
  return false
}
function pointOnSegment([x, y], [a, b], [c, d]) {
  const cross = (x - a) * (d - b) - (y - b) * (c - a)
  return Math.abs(cross) <= 1e-10 && x >= Math.min(a, c) - 1e-10 && x <= Math.max(a, c) + 1e-10 && y >= Math.min(b, d) - 1e-10 && y <= Math.max(b, d) + 1e-10
}
function inRing(point, ring) {
  let inside = false
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const a = ring[j], b = ring[i]
    if (pointOnSegment(point, a, b)) return true
    if ((b[1] > point[1]) !== (a[1] > point[1]) && point[0] < ((a[0] - b[0]) * (point[1] - b[1])) / (a[1] - b[1]) + b[0]) inside = !inside
  }
  return inside
}
function inPolygon(point, polygon) { return inRing(point, polygon[0]) && !polygon.slice(1).some((ring) => inRing(point, ring)) }
export function geometryCovers(geometry, point) {
  if (!validGeometry(geometry) || !finitePoint(point)) return false
  return geometry.type === 'Polygon' ? inPolygon(point, geometry.coordinates) : geometry.coordinates.some((polygon) => inPolygon(point, polygon))
}

export function loadWeatherSpatialIndex(dataDir) {
  if (!dataDir) fail('WEATHER_DATA_DIR 未配置')
  const root = path.resolve(dataDir), index = readJson(path.join(root, 'weather', 'index-v2.json'))
  if (index?.schemaVersion !== 2 || index?.provinceCode !== '330000' || !Array.isArray(index.nodes) || index.nodes.length === 0) fail('天气空间索引版本无效')
  const rawNodes = new Map()
  for (const raw of index.nodes) {
    const code = typeof raw?.code === 'string' ? raw.code : '', children = raw?.childrenCodes
    if (!/^33\d{4,10}$/.test(code) || rawNodes.has(code) || !LEVELS.includes(raw?.level) || typeof raw?.name !== 'string' || !raw.name.trim() || !finitePoint(raw.representativePoint) || !Array.isArray(children) || children.some((child) => typeof child !== 'string') || new Set(children).size !== children.length) fail('天气空间索引节点无效')
    rawNodes.set(code, raw)
  }
  const roots = [...rawNodes.values()].filter((node) => node.parentCode === null)
  if (roots.length !== 1 || roots[0].code !== '330000' || roots[0].level !== 'province') fail('天气空间索引根节点无效')
  for (const node of rawNodes.values()) {
    const level = LEVELS.indexOf(node.level)
    if (level > 0) {
      const parent = rawNodes.get(node.parentCode)
      if (!parent || LEVELS.indexOf(parent.level) !== level - 1 || !parent.childrenCodes.includes(node.code)) fail('天气空间索引父子关系无效')
    }
    for (const childCode of node.childrenCodes) {
      const child = rawNodes.get(childCode)
      if (!child || child.parentCode !== node.code) fail('天气空间索引子节点无效')
    }
  }
  const visited = new Set(), stack = ['330000']
  while (stack.length) { const code = stack.pop(); if (visited.has(code)) fail('天气空间索引包含环'); visited.add(code); stack.push(...rawNodes.get(code).childrenCodes) }
  if (visited.size !== rawNodes.size) fail('天气空间索引存在不可达节点')

  const featureFiles = new Map(), verified = new Map()
  function featuresFor(relative) {
    if (!featureFiles.has(relative)) {
      const collection = readJson(safePath(root, relative))
      if (collection?.type !== 'FeatureCollection' || !Array.isArray(collection.features)) fail('天气边界文件无效')
      const features = new Map()
      for (const feature of collection.features) {
        const code = typeof feature?.properties?.code === 'string' ? feature.properties.code : String(feature?.properties?.code ?? '')
        const name = typeof feature?.properties?.name === 'string' ? feature.properties.name.trim() : ''
        if (!code || !name || features.has(code) || !validGeometry(feature?.geometry)) fail('天气边界要素无效')
        features.set(code, { name, geometry: feature.geometry })
      }
      featureFiles.set(relative, features)
    }
    return featureFiles.get(relative)
  }
  for (const raw of rawNodes.values()) {
    const relative = raw.boundary?.path
    if (raw.boundary?.featureCode !== raw.code) fail('天气边界索引不一致')
    const feature = featuresFor(relative).get(raw.code)
    if (!feature || feature.name !== raw.name || !geometryCovers(feature.geometry, raw.representativePoint)) fail('天气索引名称、代表点与边界不一致')
    verified.set(raw.code, Object.freeze({ code: raw.code, name: raw.name, level: raw.level, parentCode: raw.parentCode, childrenCodes: Object.freeze([...raw.childrenCodes]), representativePoint: Object.freeze([...raw.representativePoint]), geometry: feature.geometry }))
  }
  const rootNode = verified.get('330000')
  for (const node of verified.values()) {
    if (!geometryCovers(rootNode.geometry, node.representativePoint)) fail('天气代表点越出浙江省界')
    let parent = node.parentCode ? verified.get(node.parentCode) : null
    while (parent) {
      if (!geometryCovers(parent.geometry, node.representativePoint)) fail('天气代表点越出父级行政链')
      parent = parent.parentCode ? verified.get(parent.parentCode) : null
    }
  }
  return Object.freeze({
    get(code, level) { const node = verified.get(code); if (!node || node.level !== level) throw new WeatherSpatialError('outside', '行政代码不属于当前层级或浙江省'); return node },
    covers(node, lon, lat) {
      if (verified.get(node?.code) !== node) return false
      const point = [lon, lat]
      if (!geometryCovers(rootNode.geometry, point)) return false
      let current = node
      while (current) {
        if (!geometryCovers(current.geometry, point)) return false
        current = current.parentCode ? verified.get(current.parentCode) : null
      }
      return true
    },
    province() { return rootNode },
    findAlarmNode(code) {
      const node = verified.get(code)
      return node && ['province', 'city', 'county'].includes(node.level) ? node : null
    },
    alertNodes(node) { if (verified.get(node?.code) !== node) fail('天气空间节点未验证'); if (node.level === 'province' || node.level === 'city') return node.childrenCodes.map((code) => verified.get(code)); return node.level === 'county' ? [node] : [] },
  })
}

File: scripts/weather_spatial_index.py
# -*- coding: utf-8 -*-
"""生成并校验天气服务使用的浙江五级可信空间索引。"""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path, PurePosixPath
from typing import Any

from shapely.geometry import Point, shape

SCHEMA_VERSION = 1
INDEX_RELATIVE_PATH = Path("weather") / "index-v1.json"
LEVELS = ("province", "city", "county", "township", "village")


class WeatherSpatialIndexError(ValueError):
    """天气空间索引或其可信边界不完整、不一致。"""


def _read_json(path: Path) -> dict[str, Any]:
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise WeatherSpatialIndexError(f"无法读取 JSON：{path}") from exc
    if not isinstance(value, dict):
        raise WeatherSpatialIndexError(f"JSON 根节点必须是对象：{path}")
    return value


def _safe_data_path(data_dir: Path, relative_path: str) -> Path:
    posix_path = PurePosixPath(relative_path)
    if posix_path.is_absolute() or ".." in posix_path.parts:
        raise WeatherSpatialIndexError(f"边界引用不是安全相对路径：{relative_path}")
    resolved_data = data_dir.resolve()
    resolved_path = (data_dir / Path(*posix_path.parts)).resolve()
    if resolved_path != resolved_data and resolved_data not in resolved_path.parents:
        raise WeatherSpatialIndexError(f"边界引用越出数据目录：{relative_path}")
    return resolved_path


def _load_feature_map(data_dir: Path, relative_path: str) -> dict[str, dict[str, Any]]:
    path = _safe_data_path(data_dir, relative_path)
    collection = _read_json(path)
    if collection.get("type") != "FeatureCollection" or not isinstance(collection.get("features"), list):
        raise WeatherSpatialIndexError(f"边界文件不是 FeatureCollection：{relative_path}")

    features: dict[str, dict[str, Any]] = {}
    for feature in collection["features"]:
        if not isinstance(feature, dict) or feature.get("type") != "Feature":
            raise WeatherSpatialIndexError(f"边界文件包含无效 Feature：{relative_path}")
        properties = feature.get("properties")
        code = str(properties.get("code", "")).strip() if isinstance(properties, dict) else ""
        if not code or code in features:
            raise WeatherSpatialIndexError(f"边界文件包含空或重复行政代码：{relative_path}")
        features[code] = feature
    return features


def _trusted_geometry(feature: dict[str, Any], reference: str):
    try:
        geometry = shape(feature.get("geometry"))
    except Exception as exc:
        raise WeatherSpatialIndexError(f"边界几何无法解析：{reference}") from exc
    if geometry.geom_type not in {"Polygon", "MultiPolygon"}:
        raise WeatherSpatialIndexError(f"边界几何必须是面：{reference}")
    if geometry.is_empty or not geometry.is_valid:
        raise WeatherSpatialIndexError(f"边界几何为空或无效：{reference}")
    return geometry


def _representative_point(geometry, reference: str) -> list[float]:
    point = geometry.representative_point()
    lon, lat = float(point.x), float(point.y)
    if not math.isfinite(lon) or not math.isfinite(lat) or not geometry.covers(Point(lon, lat)):
        raise WeatherSpatialIndexError(f"无法生成面内代表点：{reference}")
    return [lon, lat]


def _node(
    *, code: str, name: str, level: str, parent_code: str | None,
    children_codes: list[str], boundary_path: str, feature: dict[str, Any],
) -> dict[str, Any]:
    reference = f"{boundary_path}#{code}"
    geometry = _trusted_geometry(feature, reference)
    return {
        "code": code,
        "name": name,
        "level": level,
        "parentCode": parent_code,
        "childrenCodes": children_codes,
        "representativePoint": _representative_point(geometry, reference),
        "boundary": {"path": boundary_path, "featureCode": code},
    }


def build_weather_spatial_index(data_dir: Path) -> dict[str, Any]:
    """从 prepare-boundaries.py 的最终 GeoJSON 产物构建完整五级索引。"""
    data_dir = Path(data_dir)
    manifest = _read_json(data_dir / "manifest.json")
    province = manifest.get("province")
    cities = manifest.get("cities")
    if not isinstance(province, dict) or not isinstance(cities, list):
        raise WeatherSpatialIndexError("manifest.json 缺少 province/cities")

    province_code = str(province.get("code", "")).strip()
    if province_code != "330000":
        raise WeatherSpatialIndexError("天气空间索引只接受浙江省根节点 330000")

    feature_cache: dict[str, dict[str, dict[str, Any]]] = {}

    def features(path: str) -> dict[str, dict[str, Any]]:
        if path not in feature_cache:
            feature_cache[path] = _load_feature_map(data_dir, path)
        return feature_cache[path]

    nodes: list[dict[str, Any]] = []
    province_path = "boundary/province.geojson"
    city_path = "boundary/city/330000.geojson"
    city_codes = [str(city.get("code", "")).strip() for city in cities if isinstance(city, dict)]
    province_feature = features(province_path).get(province_code)
    if province_feature is None:
        raise WeatherSpatialIndexError("省界文件缺少 330000")
    nodes.append(_node(
        code=province_code, name=str(province.get("name", "")).strip(), level="province",
        parent_code=None, children_codes=city_codes, boundary_path=province_path,
        feature=province_feature,
    ))

    for city in cities:
        if not isinstance(city, dict) or not isinstance(city.get("counties"), list):
            raise WeatherSpatialIndexError("manifest 城市节点结构无效")
        city_code = str(city.get("code", "")).strip()
        counties = city["counties"]
        county_codes = [str(county.get("code", "")).strip() for county in counties if isinstance(county, dict)]
        city_feature = features(city_path).get(city_code)
        if city_feature is None:
            raise WeatherSpatialIndexError(f"市界文件缺少 {city_code}")
        nodes.append(_node(
            code=city_code, name=str(city.get("name", "")).strip(), level="city",
            parent_code=province_code, children_codes=county_codes, boundary_path=city_path,
            feature=city_feature,
        ))

        county_path = f"boundary/county/{city_code}.geojson"
        for county in counties:
            if not isinstance(county, dict) or not isinstance(county.get("townships"), list):
                raise WeatherSpatialIndexError(f"manifest 县节点结构无效：{city_code}")
            county_code = str(county.get("code", "")).strip()
            townships = county["townships"]
            township_codes = [str(town.get("code", "")).strip() for town in townships if isinstance(town, dict)]
            county_feature = features(county_path).get(county_code)
            if county_feature is None:
                raise WeatherSpatialIndexError(f"县界文件缺少 {county_code}")
            nodes.append(_node(
                code=county_code, name=str(county.get("name", "")).strip(), level="county",
                parent_code=city_code, children_codes=township_codes, boundary_path=county_path,
                feature=county_feature,
            ))

            township_path = f"boundary/township/{county_code}.geojson"
            for township in townships:
                if not isinstance(township, dict):
                    raise WeatherSpatialIndexError(f"manifest 乡节点结构无效：{county_code}")
                township_code = str(township.get("code", "")).strip()
                village_path = f"villages/{township_code}.geojson"
                village_file = data_dir / village_path
                village_features = features(village_path) if village_file.exists() else {}
                village_codes = sorted(village_features)
                township_feature = features(township_path).get(township_code)
                if township_feature is None:
                    raise WeatherSpatialIndexError(f"乡界文件缺少 {township_code}")
                nodes.append(_node(
                    code=township_code, name=str(township.get("name", "")).strip(), level="township",
                    parent_code=county_code, children_codes=village_codes, boundary_path=township_path,
                    feature=township_feature,
                ))
                for village_code in village_codes:
                    village_feature = village_features[village_code]
                    properties = village_feature["properties"]
                    nodes.append(_node(
                        code=village_code, name=str(properties.get("name", "")).strip(), level="village",
                        parent_code=township_code, children_codes=[], boundary_path=village_path,
                        feature=village_feature,
                    ))

    index = {
        "schemaVersion": SCHEMA_VERSION,
        "provinceCode": province_code,
        "nodes": nodes,
    }
    _validate_tree(index)
    return index


def write_weather_spatial_index(data_dir: Path) -> Path:
    data_dir = Path(data_dir)
    index = build_weather_spatial_index(data_dir)
    output = data_dir / INDEX_RELATIVE_PATH
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(index, ensure_ascii=False, indent=1) + "\n", encoding="utf-8")
    # 重读序列化结果，避免浮点序列化后代表点逸出边界。
    load_trusted_weather_spatial_index(data_dir)
    return output


def _validate_tree(index: dict[str, Any]) -> dict[str, dict[str, Any]]:
    if index.get("schemaVersion") != SCHEMA_VERSION or index.get("provinceCode") != "330000":
        raise WeatherSpatialIndexError("天气空间索引版本或省级根代码无效")
    raw_nodes = index.get("nodes")
    if not isinstance(raw_nodes, list) or not raw_nodes:
        raise WeatherSpatialIndexError("天气空间索引没有节点")

    nodes: dict[str, dict[str, Any]] = {}
    for node in raw_nodes:
        if not isinstance(node, dict):
            raise WeatherSpatialIndexError("天气空间索引包含非对象节点")
        code = str(node.get("code", "")).strip()
        level = node.get("level")
        if not code or code in nodes or level not in LEVELS:
            raise WeatherSpatialIndexError(f"天气空间索引节点代码或层级无效：{code}")
        if not isinstance(node.get("childrenCodes"), list):
            raise WeatherSpatialIndexError(f"天气空间索引 childrenCodes 无效：{code}")
        nodes[code] = node

    roots = [node for node in nodes.values() if node.get("parentCode") is None]
    if len(roots) != 1 or roots[0].get("code") != "330000" or roots[0].get("level") != "province":
        raise WeatherSpatialIndexError("天气空间索引必须且只能有浙江省根节点")

    for code, node in nodes.items():
        level_index = LEVELS.index(node["level"])
        parent_code = node.get("parentCode")
        if level_index:
            parent = nodes.get(parent_code)
            if parent is None or parent["level"] != LEVELS[level_index - 1] or code not in parent["childrenCodes"]:
                raise WeatherSpatialIndexError(f"天气空间索引父子关系无效：{code}")
        for child_code in node["childrenCodes"]:
            child = nodes.get(child_code)
            if child is None or child.get("parentCode") != code:
                raise WeatherSpatialIndexError(f"天气空间索引子节点引用无效：{code}->{child_code}")
    return nodes


def load_trusted_weather_spatial_index(data_dir: Path) -> dict[str, dict[str, Any]]:
    """Fail closed 地加载索引，并逐节点核对最终边界与代表点 covers 关系。"""
    data_dir = Path(data_dir)
    index = _read_json(data_dir / INDEX_RELATIVE_PATH)
    nodes = _validate_tree(index)
    feature_cache: dict[str, dict[str, dict[str, Any]]] = {}
    for code, node in nodes.items():
        boundary = node.get("boundary")
        point = node.get("representativePoint")
        if not isinstance(boundary, dict) or not isinstance(point, list) or len(point) != 2:
            raise WeatherSpatialIndexError(f"天气空间索引边界或代表点结构无效：{code}")
        path = boundary.get("path")
        feature_code = boundary.get("featureCode")
        if not isinstance(path, str) or feature_code != code:
            raise WeatherSpatialIndexError(f"天气空间索引边界引用无效：{code}")
        if path not in feature_cache:
            feature_cache[path] = _load_feature_map(data_dir, path)
        feature = feature_cache[path].get(code)
        if feature is None:
            raise WeatherSpatialIndexError(f"可信边界缺少索引节点：{path}#{code}")
        geometry = _trusted_geometry(feature, f"{path}#{code}")
        try:
            lon, lat = float(point[0]), float(point[1])
        except (TypeError, ValueError) as exc:
            raise WeatherSpatialIndexError(f"天气空间索引代表点不是数值：{code}") from exc
        if not math.isfinite(lon) or not math.isfinite(lat) or not geometry.covers(Point(lon, lat)):
            raise WeatherSpatialIndexError(f"代表点不在最终可信边界内：{code}")
    return nodes


def main() -> int:
    parser = argparse.ArgumentParser(description="生成并校验天气五级可信空间索引")
    parser.add_argument("--data-dir", type=Path, default=Path(__file__).resolve().parent.parent / "web/public/data")
    parser.add_argument("--validate-only", action="store_true")
    args = parser.parse_args()
    try:
        if args.validate_only:
            nodes = load_trusted_weather_spatial_index(args.data_dir)
            print(f"天气空间索引校验通过：{len(nodes)} 个节点")
        else:
            output = write_weather_spatial_index(args.data_dir)
            print(f"天气空间索引已生成并校验：{output}")
        return 0
    except WeatherSpatialIndexError as exc:
        print(f"天气空间索引失败：{exc}")
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
